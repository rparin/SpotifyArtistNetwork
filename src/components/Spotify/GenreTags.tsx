"use client";
import React, { useRef, useEffect, useState } from "react";
import { ChevronLeft, ChevronRight } from "lucide-react";
import cn from "@/utils/cn";

export default function GenreTags({
  id,
  items,
  hScrollAmt = 40,
  className,
}: {
  id: string;
  items: Array<string>;
  hScrollAmt?: number;
  className?: string;
}) {
  const ref = useRef<HTMLUListElement>(null);
  const [arrows, setArrows] = useState({ left: false, right: true });
  const centerGenre = items.length == 1 ? "pl-3 justify-center" : "";

  const genres = items.map((item, index) => {
    return (
      <li
        key={`${id}-Genre-${index}`}
        className={cn("rounded-xl bg-white px-3 dark:bg-black", className)}>
        {item}
      </li>
    );
  });

  const hScroll = (amt: number) => {
    if (!ref.current) return;
    ref.current.scrollLeft += amt;
  };

  useEffect(() => {
    const refCurrent = ref.current;
    const handleScroll = () => {
      if (!refCurrent) return;
      const arrowOffset = 15;

      // Disable left arrow when no items to the left
      if (refCurrent.scrollLeft <= arrowOffset) {
        setArrows({
          left: false,
          right: true,
        });
      }

      // Show both arrows when in the middle of the list
      else if (
        refCurrent.scrollLeft <=
        refCurrent.scrollWidth - ref.current.offsetWidth
      ) {
        setArrows({
          left: true,
          right: true,
        });
      }

      // Disable right arrow when at the end of the list
      if (
        refCurrent.scrollLeft >=
        refCurrent.scrollWidth - refCurrent.offsetWidth - arrowOffset
      ) {
        setArrows({
          left: true,
          right: false,
        });
      }
    };

    if (refCurrent) {
      refCurrent.addEventListener("scroll", handleScroll);
    }

    // cleanup function to remove event listener on unmount
    return () => {
      if (refCurrent) {
        refCurrent.removeEventListener("scroll", handleScroll);
      }
    };
  }, []);

  const getButtons = () => {
    if (items.length > 1) {
      return (
        <>
          <button
            onClick={() => hScroll(-hScrollAmt)}
            className="invisible absolute left-[-0.12rem] top-[-2px] group-hover:visible">
            {arrows.left && <ChevronLeft />}
          </button>
          <button
            onClick={() => hScroll(hScrollAmt)}
            className="invisible absolute right-[-0.12rem] top-[-2px] group-hover:visible">
            {arrows.right && <ChevronRight />}
          </button>
        </>
      );
    }
  };

  const getGenreTags = () => {
    // No genres to show
    if (items.length == 0) {
      return <span className="py-3"></span>;
    }

    // Display list of genres
    return (
      <div className="no-scrollbar group relative w-full overflow-y-hidden px-4">
        {/* Display arrow buttons if more than 1 genre */}
        {getButtons()}
        <ul
          ref={ref}
          className={
            "no-scrollbar horizontal-mask flex gap-2 overflow-x-scroll whitespace-nowrap px-3 pl-7 text-sm text-black dark:text-white " +
            centerGenre
          }>
          {genres}
        </ul>
      </div>
    );
  };

  return <>{getGenreTags()}</>;
}
