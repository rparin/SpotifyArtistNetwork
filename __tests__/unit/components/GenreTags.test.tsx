import { render } from "@testing-library/react";
import { axe, toHaveNoViolations } from "jest-axe";
import GenreTags from "@/components/Spotify/GenreTags";

const PROPS = {
  id: "0bAsR2unSRpn6BQPEnNlZm",
  items: ["pop", "rock"],
};

test("it renders component unchanged", () => {
  const { container } = render(<GenreTags {...PROPS} />);
  expect(container).toMatchSnapshot();
});

describe("testing Genre Tags Accessibility", () => {
  expect.extend(toHaveNoViolations);

  test("it should have no accessibility violations", async () => {
    const { container } = render(<GenreTags {...PROPS} />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
