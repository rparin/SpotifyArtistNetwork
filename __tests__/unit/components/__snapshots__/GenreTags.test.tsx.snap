// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders component unchanged 1`] = `
<div>
  <div
    class="no-scrollbar group relative w-full overflow-y-hidden px-4"
  >
    <button
      aria-label="Genre Tags: Scroll Left"
      class="invisible absolute left-[-0.12rem] top-[-2px] group-focus-within:visible group-hover:visible"
    />
    <button
      aria-label="Genre Tags: Scroll Right"
      class="invisible absolute right-[-0.12rem] top-[-2px] group-focus-within:visible group-hover:visible"
    >
      <svg
        class="lucide lucide-chevron-right"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m9 18 6-6-6-6"
        />
      </svg>
    </button>
    <ul
      class="no-scrollbar horizontal-mask flex gap-2 overflow-x-scroll whitespace-nowrap px-3 pl-7 text-sm text-black dark:text-white "
      tabindex="0"
    >
      <li
        class="rounded-xl bg-white px-3 dark:bg-black"
      >
        pop
      </li>
      <li
        class="rounded-xl bg-white px-3 dark:bg-black"
      >
        rock
      </li>
    </ul>
  </div>
</div>
`;
